TestIPServe fails sporadically

TestIPServe fails sporadically in Travis-CI:

    === RUN TestIPServe-2
    --- FAIL: TestIPServe-2 (0.00s)
    	server_test.go:32: Error in connecting to IPAM server
    FAIL
    exit status 1
    FAIL	github.com/openshift/openshift-sdn/pkg/netutils/server	0.005s

eg, https://travis-ci.org/openshift/openshift-sdn/builds/76176233, https://travis-ci.org/openshift/openshift-sdn/builds/76204528, https://travis-ci.org/openshift/openshift-sdn/builds/76209682

It's always the second run that fails... maybe something isn't getting cleaned up correctly after the first one?

Just failed again: https://travis-ci.org/openshift/openshift-sdn/jobs/79471475:

    === RUN TestIPServe-2
    --- FAIL: TestIPServe-2 (0.00s)
    	server_test.go:32: Error in connecting to IPAM server: %v Get http://127.0.0.1:9080/netutils/ip: dial tcp 127.0.0.1:9080: connection refused

(I already fixed the "%v" in git.)

--- a/ovssubnet/controller/kube/kube.go
+++ b/ovssubnet/controller/kube/kube.go
@@ -66,7 +66,6 @@ func (c *FlowController) manageLocalIpam(ipnet *net.IPNet) error {
 		return err
 	}
 	f.Close()
-	// listen and serve does not return the control
 	netutils_server.ListenAndServeNetutilServer(ipam, net.ParseIP(ipamHost), ipamPort, nil)
 	return nil
 }

--- a/pkg/netutils/server/server.go
+++ b/pkg/netutils/server/server.go
@@ -29,21 +29,27 @@ type IpamInterface interface {
 }
 
 // ListenAndServeNetutilServer initializes a server to respond to HTTP network requests on the ipam interface
-func ListenAndServeNetutilServer(ipam IpamInterface, address net.IP, port uint, tlsOptions *TLSOptions) {
+func ListenAndServeNetutilServer(ipam IpamInterface, address net.IP, port uint, tlsOptions *TLSOptions) error {
 	handler := NewServer(ipam)
+	addr := net.JoinHostPort(address.String(), strconv.FormatUint(uint64(port), 10))
 	s := &http.Server{
-		Addr:           net.JoinHostPort(address.String(), strconv.FormatUint(uint64(port), 10)),
 		Handler:        handler,
 		ReadTimeout:    5 * time.Minute,
 		WriteTimeout:   5 * time.Minute,
 		MaxHeaderBytes: 1 << 20,
 	}
+	var listener net.Listener
+	var err error
 	if tlsOptions != nil {
-		s.TLSConfig = tlsOptions.Config
-		s.ListenAndServeTLS(tlsOptions.CertFile, tlsOptions.KeyFile)
+		listener, err = tls.Listen("tcp", addr, tlsOptions.Config)
 	} else {
-		s.ListenAndServe()
+		listener, err = net.Listen("tcp", addr)
 	}
+	if err != nil {
+		return err
+	}
+	go s.Serve(listener)
+	return nil
 }
 
 // NewServer initializes and configures the netutils_server.Server object to handle HTTP requests.

--- a/pkg/netutils/server/server_test.go
+++ b/pkg/netutils/server/server_test.go
@@ -45,7 +45,7 @@ func TestIPServe(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Error while initializing IPAM: %v", err)
 	}
-	go ListenAndServeNetutilServer(ipam, net.ParseIP("127.0.0.1"), 9080, nil)
+	ListenAndServeNetutilServer(ipam, net.ParseIP("127.0.0.1"), 9080, nil)
 
 	// get, get, delete, get
 	ip := getIP(t)

